### DISCUSSION TITLE
Depending on a crate defined in a flake

### DISCUSSION THREAD

**Discussion by szg251** [90-008/nix-cargo-integration#139] [General]
I have mutiple private packages (not published on any package registry). Each one of them uses `nci` and I want crate-a to depend on crate-b. Is this possible?

**Comment by 90-008** [comment]
I'm not sure I understand the question fully, I'd assume you just depend on a crate the regular way by adding it to Cargo dependencies? I don't think NCI or not matters here.

**Reply by szg251** [reply]
The key difference is that it's not published on a package registry, so cargo would have to find it from the nix derivation. I'm new to Rust, I know this is possible with Haskell.

**Reply by szg251** [reply]
So crate-b would be a nix flake input to crate-a. I know this is a grey area, some say the native build environment (in this case cargo) should always be single source of truth, and nix kind of optional.
I'll do some research today, and see what's best suited for this scenario.
Btw, thanks for the quick responses!

**Reply by 90-008** [reply]
Usually what I would do with private crates is just clone them locally onto a separate folder and then add them as local path dependencies in Cargo. Though this would cause issues with NCI because it would be outside of the flake's context and the path wouldn't be accessible. Alternatively you could use a private git repository and add the necessary authentication to Nix to make it work, this could possibly work but I haven't tried it before with Cargo.

For the local path solution it would be possible to just point to the flake and then patch the vendor folder that Cargo will use when building, just needs some support in NCI for it. I can work on this sometime. Submodules are also one possible solution but they don't work well with flakes (and IMO they suck in general) so I'd avoid them.

**Comment by 90-008** [comment]
i don't think i can support this in nci since it doesn't do any building logic itself (dream2nix does but nci is still a cargo focused project, this would require nci to have some sort of build graph etc feature and call rustc via nixpkgs' buildRustCrate to link crates manually), so will have to close

**Comment by szg251** [comment]
@yusdacra Thank you for your help. What I ended up doing is I symlinked the dependencies into a subfolder, and used something like `{ path = ./mysubfolder/crate-a }`, not ideal, but does the job for now.
If I have the time (which I don't) I might play around with a custom package repository built by nix. That would be the ideal solution for my usecase.

**Reply by 90-008** [reply]
interesting, well, glad to know it at least somehow works! i'd like to add a builder that uses `buildRustCrate` to dream2nix eventually, it would be nice (crate-level granularity like crate2nix).
